### PR TITLE
Compute shrink-to-fit width for flex blocks

### PR DESCRIPTION
**Pull request by alml**
Refs https://github.com/mikke89/RmlUi/issues/552

### CODE REVIEW

**Comment by mikke89** [comment]
Thanks a lot for the pull request, enhancements to this part of the layout engine is very much welcome.

Can you expand a bit on how this relates to the specific issues in #552? Before and after comparisons would be helpful. If you are still working on this, let me know when you feel it is ready for review.

**Comment by alml** [comment]
The implementation passes my own tests. In particular, I was able to correctly render a pretty complex UI in my MMO game. I made a few corrections to the PR recently after I discovered a couple more corner cases. The major one was an assertion in the flex layout engine where it incorrectly estimated size of nested elements when both element and container had an auto width. Examples that cover those corner cases are available below.

From the correctness standpoint, you can review and even merge this PR right away. I'm pretty confident that it won't break existing applications and can only make it better for auto-sized flex elements.

However, from performance standpoint, it's terribly slow. Existing applications may be hit by the combinatorial explosion that you highlighted in https://github.com/mikke89/RmlUi/issues/552#issuecomment-1869817300. The engine tries to compute the same layouts of nested elements with the same parameters over and over again. I think if we do some memoization of computed layouts to return the same result if called with the same parameters, it would immediately yield enormous performance benefits reducing algorithmic complexity from O(2^N) to O(N).

I would appreciate it if you could review the code as is (maybe move it to another branch) and then perhaps help with caching/memoization, because it's going to be a pretty deep change to the layout engine.

If you have other ideas, I'd love to hear them.

# Case 1

```
<rml>
	<head>
		<style>
			* {
				box-sizing: border-box;
		    	        font-family: Play;
			}

			.outer {
			        display: flex;
			        border: 1px red;
                                padding: 50px;
			}

                       .inner {
                                border: 1px blue;
                                padding: 50px;
                       }
		</style>
	</head>
	<body>
                <div class="outer"><div class="inner">Block 1</div></div>
	</body>
</rml>
```

Before:
image

After:
image

# Case 2

```
<rml>
    <head>
        <style>
            * {
                box-sizing: border-box;
                font-family: Play;
            }

            .block {
                display: block;
                width: 50px;
                height: 50px;
                margin: 10px;
            }

            .red {
                background-color: red;
            }

            .green {
                background-color: green;
            }

            .blue {
                background-color: blue;
            }

            .outer {
                display: flex;
                flex-direction: row;
            }

            .inner {
                display: flex;
                flex-direction: column;
            }
        </style>
    </head>
    <body>
        <div class="outer">
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
        </div>
    </body>
</rml>
```

Before:
image

After:
image

# Case 3

```
<rml>
    <head>
        <style>
            * {
                box-sizing: border-box;
                font-family: Play;
            }

            .block {
                display: block;
                width: 50px;
                height: 50px;
                margin: 10px;
            }

            .red {
                background-color: red;
            }

            .green {
                background-color: green;
            }

            .blue {
                background-color: blue;
            }

            .outer {
                display: flex;
                flex-direction: column;
            }

            .inner {
                display: flex;
                flex-direction: row;
            }
        </style>
    </head>
    <body>
        <div class="outer">
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
            <div class="inner">
                <div class="block red"></div>
                <div class="block green"></div>
                <div class="block blue"></div>
            </div>
        </div>
    </body>
</rml>
```

Before:
image

After:
image

# Case 4

```
<rml>
	<body style="display: flex; flex-direction: column">
	        <div style="padding: 42px; background: blue; border: 5px #00ffff">
	        	<div style="display: flex; background: green; padding: 10px; border: 5px #ff00ff">
				<div style="width: 10px; height: 10px; background: red; border: 5px #ffff00">
				</div>
			</div>
        	</div>
	</body>
</rml>
```

Before:
image

And a failed assertion in the console:
```
2024-01-05T20:27:52.7760615+00:00 [ERROR] RML: RMLUI_ASSERT(containing_block.x >= 0.f)
external/rmlui~override/Source/Core/Layout/FlexFormattingContext.cpp:53
```

After:
image

# Case 5

```
<rml>
    <head>
        <style>
            * { align-items: flex-start; }
        </style>
    </head>
    <body id="o" style="display: flex; flex-direction: row; width: 100%; height: 100%">
        <div id="a" style="display: flex; flex-direction: row; border: 5px yellow">
            <div id="b" style="display: flex; flex-direction: row; border: 5px red; width: 100px; height: 100px;">
                <div id="c" style="display: flex; flex-direction: row; border: 5px green">
                    <div id="d" style="display: flex; flex-direction: row; border: 5px blue; width: 50px; height: 50px;">
                    </div>
                </div>
            </div>
        </div>
    </body>
</rml>
```

Before:
image

After:
image

# Case 6

```
<rml>
    <head>
        <style>
            * { align-items: flex-end; }
        </style>
    </head>
    <body id="o" style="display: flex; flex-direction: column; width: 100%; height: 100%">
        <div id="a" style="display: flex; flex-direction: column; border: 5px yellow">
            <div id="b" style="display: flex; flex-direction: column; border: 5px red; width: 100px; height: 100px;">
                <div id="c" style="display: flex; flex-direction: column; border: 5px green">
                    <div id="d" style="display: flex; flex-direction: column; border: 5px blue; width: 50px; height: 50px;">
                    </div>
                </div>
            </div>
        </div>
    </body>
</rml>
```

Before:
image

After:
image

**Comment by alml** [comment]
Actually, you know what? it turned out that caching shrink-to-fit width is pretty simple. Check it out - https://github.com/mikke89/RmlUi/pull/559/commits/d9a4627e41e532b4ab62b15aec7949bf6653687d. Layout performance immediately went back to normal.

File: Source/Core/Layout/ContainerBox.cpp [hunk 1]
```diff
@@ -274,7 +275,8 @@ float FlexContainer::GetShrinkToFitWidth() const
 	if (element->GetComputedValues().width().type == Style::Width::Type::Length)
 		return box.GetSize().x;
 
-	return 0.0f;
+	// Infer shrink-to-fit width from the intrinsic width of the element.
```
Comment (mikke89) [inline Source/Core/Layout/ContainerBox.cpp:278]: The comment at the top of this function could be revised, now that we support this :)

File: Source/Core/Layout/LayoutDetails.cpp [hunk 2]
```diff
@@ -275,9 +275,13 @@ float LayoutDetails::GetShrinkToFitWidth(Element* element, Vector2f containing_b
 	RootBox root(Math::Max(containing_block, Vector2f(0.f)));
 	UniquePtr<LayoutBox> layout_box = FormattingContext::FormatIndependent(&root, element, &box, FormattingContextType::Block);
 
-	const float available_width = Math::Max(0.f, containing_block.x - box.GetSizeAcross(BoxDirection::Horizontal, BoxArea::Margin, BoxArea::Padding));
-
-	return Math::Min(available_width, layout_box->GetShrinkToFitWidth());
+	float shrink_to_fit_width = layout_box->GetShrinkToFitWidth();
+	if (containing_block.x >= 0) {
```
Comment (mikke89) [inline Source/Core/Layout/LayoutDetails.cpp:285]: The formatting is a bit off, could you run the changed files through clang-format?

**Review by mikke89 — COMMENTED** [review]
This approach looks simple and effective, nice! :) The spec goes on about handling different flex factors and such, but a simpler approach could be more suitable for us. Regardless, this is non-breaking behavior, as we didn't support shrink-to-fit here previously. And then we could improve later if we find a case for that.

I believe we might have some issues with handling percentage-based sizing here (and generally in shrink-to-fit contexts?), which is something we might want to improve later. But this probably requires more thorough layout engine-wide changes implementing stronger typing for indefinite sizing and sizing constraints.

I ran this through all of our own and generated layout tests to confirm we didn't break anything. Most tests are equal to before, only two CSS flex tests had been changed. The changes being improvements - now both of them matches their reference! Attaching screenshots of the changes at the bottom here .

Your examples here look great too, appreciate all the details. I wonder if you could add some, or all of these to the VisualTests? That way we ensure they don't break in the future.

I only have some minor comments for this first commit, otherwise I'm mostly ready to merge this.

When it comes to the caching commit, this is something I have very much been interested in looking at for a while. However, that is also a very big change (code is small but implications are large), and requires a lot more consideration. We should move that to a new PR and discuss things thoroughly.

---

Yellow frame is before, document to the right is reference.

test

test2

File: Source/Core/Layout/FlexFormattingContext.cpp [hunk 3]
```diff
@@ -315,10 +336,11 @@ void FlexFormattingContext::Format(Vector2f& flex_resulting_content_size, Vector
 			RMLUI_ASSERT(initial_box_size.y < 0.f);
 
 			Box format_box = item.box;
-			if (initial_box_size.x < 0.f)
+			if (initial_box_size.x < 0.f && flex_available_content_size.x >= 0.f)
```
[the remark this answers is not in the document]
Comment (alml) [reply Source/Core/Layout/FlexFormattingContext.cpp:339]: Frankly I don't fully understand this code and can't exhaustively list all cases when it can happen. It definitely happened in shrink-to-fit context, and there it meant infinite max-content box. What I can say for sure that if a negative width makes its way to SetContent, then it will definitely fail an assertion later on. See my original comment. I was able to trip that assertion even before my changes, so I think this change is an improvement. Whether it's universally correct, I don't know.

File: Source/Core/Layout/FlexFormattingContext.cpp [hunk 4]
```diff
@@ -315,10 +336,11 @@ void FlexFormattingContext::Format(Vector2f& flex_resulting_content_size, Vector
...
 				format_box.SetContent(Vector2f(flex_available_content_size.x - item.cross.sum_edges, initial_box_size.y));
 
-			FormattingContext::FormatIndependent(flex_container_box, element, &format_box, FormattingContextType::Block);
+			FormattingContext::FormatIndependent(flex_container_box, element,
+				(format_box.GetSize().x >= 0 ? &format_box : nullptr), FormattingContextType::Block);
```
[the remark this answers is not in the document]
Comment (alml) [reply Source/Core/Layout/FlexFormattingContext.cpp:343]: Could you explain what you mean here? This condition is different from the previous one.

* `flex_available_content_size.x >= 0.f` verifies that the available space is finite. If it's infinite, we are not even trying to override the box and use something default.
* After that we update the box by setting new content size.
* I suppose as a result `format_box.Size()` can go negative (e.g. if `sum_edges` is very big)
* To make sure I don't feed invalid `overide_initial_box` to FormatIndependent, I make this check.

Please advise how to proceed from here.

File: Source/Core/Layout/LayoutDetails.cpp [hunk 5]
```diff
@@ -256,16 +256,16 @@ float LayoutDetails::GetShrinkToFitWidth(Element* element, Vector2f containing_b
...
-		display == Style::Display::InlineTable)
+	if (display == Style::Display::InlineFlex || display == Style::Display::Table || display == Style::Display::InlineTable) {
 		return 0.f;
+	}
 
 	// Use a large size for the box content width, so that it is practically unconstrained. This makes the formatting
 	// procedure act as if under a maximum content constraint. Children with percentage sizing values may be scaled
 	// based on this width (such as 'width' or 'margin'), if so, the layout is considered undefined like in CSS 2.
-	const float max_content_constraint_width = containing_block.x + 1000.f;
+	const float max_content_constraint_width = containing_block.x + 10000.f;
```
[the remark this answers is not in the document]
Comment (alml) [reply Source/Core/Layout/LayoutDetails.cpp:274]: With the old state, even if my screen resolution is more than 1000, the layout engine wraps long text at 1000 pixels, so it can never consume the whole screen. I increased it to 10000 assuming that there screens with higher resolution are not common.

File: Source/Core/Layout/FlexFormattingContext.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikke89) [reply Source/Core/Layout/FlexFormattingContext.cpp]: Yeah, you're right, there are many conditions where this could be non-negative, regardless of available space.

File: Source/Core/Layout/FlexFormattingContext.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikke89) [reply Source/Core/Layout/FlexFormattingContext.cpp:340]: Yeah, I believe generally this is considered indefinite, which in the context of available space would mean infinite. The spec goes into [a lot more detail](https://www.w3.org/TR/css-flexbox-1/#flex-base-size) for these calculations, but his should definitely be an improvement.

I think it is fine as it is written here.

File: Source/Core/Layout/LayoutDetails.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikke89) [reply Source/Core/Layout/LayoutDetails.cpp:269]: Makes sense, thanks!

**Comment by mikke89** [comment]
I'm happy with these changes now, thanks! The only thing I am hoping for is to add some or all of these cases to the visual tests.

**Comment by Paril** [comment]
Just out of curiosity, does this change allow `inline-flex` to be removed from the various checks and use the shrink-to-width now, or is that a different can of worms?

**Comment by alml** [comment]
Hey @Paril, I did not test if my change is sufficient to support `inline-flex`. Please look at https://github.com/mikke89/RmlUi/pull/559/files#diff-a61d267ac4ed1715193d50f64ac3b0458e22790a8a270be7a7e9c89d5970e0f4R261 - try to remove `InlineFlex` check from the condition and see what happens. If it works correctly, I'm happy to update my PR.

@mikke89 Could you merge it as is and help with visual tests as a follow-up? My environment is non-standard - I use bazel to build RmlUi instead of cmake, and don't have much time figuring out how to install all the necessary dependencies on Windows.

**Comment by Paril** [comment]
> Hey @Paril, I did not test if my change is sufficient to support `inline-flex`. Please look at https://github.com/mikke89/RmlUi/pull/559/files#diff-a61d267ac4ed1715193d50f64ac3b0458e22790a8a270be7a7e9c89d5970e0f4R261 - try to remove `InlineFlex` check from the condition and see what happens. If it works correctly, I'm happy to update my PR.

Surprisingly, I did do this, and it appeared to work - but I don't know enough about the codebase to know if that's the only thing that prevents it from working properly, heh. See https://github.com/mikke89/RmlUi/issues/577#issuecomment-1897740928

I might be able to help with the visual tests, since I pretty much exclusively use flexbox in any web development that I do.

**Comment by alml** [comment]
I don't think it's that surprising. IIUC `inline-flex` has only minor differences compared to `flex` - basically how the resulting block is laid out, not how its content is laid out (which this PR is about). Since you confirmed that it worked for you, I removed the check. Anyway it's going to be an improved compared to what RmlUi had before, so @mikke89 seems to be open to such changes.

**Comment by Paril** [comment]
Keep in mind it was two places - I had to fix this line as well:
https://github.com/mikke89/RmlUi/blob/d0de094ba321e7bdd975f4c90933ad553f00ced8/Source/Core/Layout/LayoutDetails.cpp#L453

**Comment by Paril** [comment]
Cool, let me re-merge and see how it looks. I'll edit this comment when I test it.

**Comment by mikke89** [comment]
Great, thanks a lot! This is a very nice addition.

It definitely makes sense to enable this behavior for inline-flex too.

I'll go ahead and add those visual tests.

**Comment by mikke89** [comment]
Visual tests added here: ab20b36172ff8f9fd3df265bd89d1069ee3056e1

Also, added a benchmark: 37d0001d0923d6c37a36e32edb53a980d8ce4a3a

This PR didn't affect the numbers for our existing benchmarks. After adding some new shrink-to-fit cases, naturally we see a slow-down in these cases, compared to the before the changes in this PR, where we just didn't really format it.

Before:

| relative |               ns/op |                op/s |    err% |     total | Flexbox shrink-to-fit
|---------:|--------------------:|--------------------:|--------:|----------:|:----------------------
|   100.0% |            1,891.67 |          528,634.36 |    1.5% |      0.00 | `Reference`
|     8.4% |           22,620.00 |           44,208.66 |    1.3% |      0.00 | `Basic shrink-to-fit`
|     1.2% |          156,100.00 |            6,406.15 |    0.6% |      0.00 | `Nested shrink-to-fit`

After:

| relative |               ns/op |                op/s |    err% |     total | Flexbox shrink-to-fit
|---------:|--------------------:|--------------------:|--------:|----------:|:----------------------
|   100.0% |            1,894.64 |          527,803.96 |    0.4% |      0.00 | `Reference`
|     6.7% |           28,375.00 |           35,242.29 |    1.5% |      0.00 | `Basic shrink-to-fit`
|     0.1% |        2,911,200.00 |              343.50 |    0.3% |      0.03 | `Nested shrink-to-fit`

As discussed earlier, the nested case is definitely something to be aware of.
